### PR TITLE
294 refactor product to component

### LINKED ISSUE #294: Refactor - Product to Component

### PR DESCRIPTION
**Pull request by timler**
Closes #294
Depends on  #318

### CODE REVIEW

**Comment by timler** [comment]
@danfuterman This is ready to be reviewed and merged into master. I've done an update from the current master which will hopefully make the merge easier for you. I am happy to do one more merge up after #327 has been merged into master.

**Comment by danfuterman** [comment]
Thanks @timler - #327 has been merged - I'll wait until you've merged that into your working branch before I review.

**Comment by timler** [comment]
@danfuterman I did the merge - just 3 files in conflict (Tests). I will go onto the front-end now.

**Comment by danfuterman** [comment]
@timler - Initial review looking good. UsageRepository and UsageRepositoryTest still make use of a parameter `productNumber`, which should probably be replaced with `componentId` or `componentIdentificationNumber` - need to see which of those fields is intended to be used here.

**Comment by timler** [comment]
@danfuterman yes, I know the productNumber is still there. I left it because it's actually a bug I found when I was doing testing. I didn't think it was right to do bug fixing at the same time as refactoring, but happy to fix it if you'd like me to.

**Comment by danfuterman** [comment]
@timler - Can you clarify the issue/bug? Would there be an impact on any database fields? It would be better to isolate all refactor and DB changes to a single branch/PR - so I'd suggest either the bug is handled in a separate branch and those changes are merged in here (this would delay the merging of this PR), or make temporary holder changes to this PR (so that `product` is no longer used), and then fix the bug afterwards. But I'm not sure if that's possible without understanding the bug.

**Comment by timler** [comment]
@danfuterman I added it in my summary comment on https://github.com/jembi/bsis/issues/318.

Under "Issues discovered" please see:

ProductRepository: findProduct(String) has a bug in the HSQL with property productNumber 

UsageRepository: findProductUsage, findUsageByProductNumber(String), deleteUsage(String) and findAnyUsageMatching(String, String, String, Map) have errors in the HSQL with productNumber 

Basically it looks like there was an earlier refactor and these methods were never tested. There is no property "productNumber" in Product or Component.

The failing tests were marked as ignored.

Please advise how you'd like me to handle these.

**Comment by danfuterman** [comment]
It should be fine to replace productNumber with `component_id` for now.

**Comment by timler** [comment]
@danfuterman I did the fix, please see revision 1c1dc25.

File: ddl_mysql.sql [hunk 1]
```diff
@@ -650,13 +650,13 @@
         expiresAfterUnits varchar(30),
         hasBloodGroup boolean,
         isDeleted boolean,
-        productType varchar(50),
-        productTypeNameShort varchar(30),
-        pediProductType_id SMALLINT,
+        componentType varchar(50),
```
Comment (danfuterman) [inline ddl_mysql.sql]: There seems to be a slight inconsistency in `ComponentType` and `ComponentType_AUD` - in `ComponentType_AUD` the field `productType` is renamed `componentType`, in `ComponentType` it's renamed to `componentTypeName`: https://github.com/jembi/bsis/pull/333/files#diff-06f9af6af60aa02f866fecb11ecbde07R615

Comment (danfuterman) [inline ddl_mysql.sql]: Seems related to the changeSet that was removed here https://github.com/jembi/bsis/pull/333/files#diff-c06a27a0eaed87176cf7cada630457efL1151 - but I think that `ComponentType_AUD. componentType` should be renamed to `componentTypeName` for consistency.

Comment (timler) [inline ddl_mysql.sql]: Thanks for picking that up @danfuterman - I will fix it now as you suggested (which is what I agree should be the correct naming). The manual DB upgrade script is fine.

**Comment by timler** [comment]
@danfuterman I've fixed the issue. This definitely highlights that the _AUD tables are not being tested.

**Comment by danfuterman** [comment]
Thanks @timler - agreed, we haven't focused at all on testing the AUD tables, and even when system testing these issues won't really be picked up, since we don't do a lot of editing of entities when testing. Definitely an area that could use a little more attention.
